### DISCUSSION TITLE
FYI: libxcb 1.15 crashes when attempting to build from source using xcb-proto 1.16.0

### DISCUSSION THREAD

**Discussion by Viqsi** [Homebrew/discussions#4726] [Tap maintenance and Homebrew development]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.5
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: b8b46501e4caa1cad75b30627c123127ab09a6a7
Last commit: 4 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 34b6c7d7e3a2e68133f32fa73078c15d884bd89c
Core tap last commit: 2 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: mvim -f --nomru -c "au VimLeave * !open -a iTerm"
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_INSTALL_FROM_API: set
Homebrew Ruby: 2.6.10 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: octa-core 64-bit nehalem
Clang: 12.0.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.4.0.0.1.1610135815
Xcode: N/A
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/SUVMMFaker.dylib

Warning: You are using macOS 10.15.
We (and Apple) do not provide support for this old version.
It is expected behaviour that some formulae will fail to build in this old version.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

### Description of issue

There's a bug in libxcb 1.15 whenever one is generating code for an enum with documented parameters (as is newly the case in composite.xml and damage.xml in xcb-proto 1.16.0); it crashes because the name being used to check for namespace collisions is incorrectly assembled when generating documentation in the header files.

Upstream fix is here:
https://gitlab.freedesktop.org/xorg/lib/libxcb/-/commit/33f3dbe3699a92e8ca18f470adac456e0b935e75

The resulting error looks like this:
```
==> Upgrading libxcb
  1.15 -> 1.15_1

==> Patching
==> ./configure --sysconfdir=/usr/local/etc --localstatedir=/usr/local/var --enable-dri3 --enable-ge --enable-xevie --enable-xprint
==> make
Last 15 lines from /Users/jvc/Library/Logs/Homebrew/libxcb/02.make:
  File "/private/tmp/libxcb-20230817-76925-ru8xg4/libxcb-1.15/src/./c_client.py", line 3198, in c_request
    _c_request_helper(self, name, void=True, regular=False)
  File "/private/tmp/libxcb-20230817-76925-ru8xg4/libxcb-1.15/src/./c_client.py", line 2273, in _c_request_helper
    if namecount[tname] > 1:
       ~~~~~~~~~^^^^^^^
KeyError: 'xcb_report_level_t'
    _c_request_helper(self, name, void=True, regular=False)
  File "/private/tmp/libxcb-20230817-76925-ru8xg4/libxcb-1.15/src/./c_client.py", line 2273, in _c_request_helper
    if namecount[tname] > 1:
       ~~~~~~~~~^^^^^^^
KeyError: 'xcb_redirect_t'
make[1]: *** [composite.c] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [damage.c] Error 1
make: *** [all-recursive] Error 1
```

Downgrading xcb-proto back to 1.15, updating libxcb to 1.16 (tarball at https://www.x.org/releases/individual/proto/xcb-proto-1.16.0.tar.xz), or adding upstream's fix as a patch (see above) fixes the issue. One of those first two options is probably best, since libxcb and xcb-proto always release revs side-by-side (and did so again for 1.16). I therefore strongly believe (but cannot currently prove) that this would affect anyone building from source regardless of MacOS version - which is why I'm bothering to bring it up here rather than just gripe to myself about being left behind. 😄

Hope this helps!

**Comment by SMillerDev** [comment]
Could you make a pull request to add the patch? Since you know the details, that seems like te best way.

**Reply by Viqsi** [reply]
Pardon the delay - I can try to do that, altho I've literally never done a pull request before. Been reading about how. :)

**Comment by chujiu** [comment]
非常感谢，我是这么解决的！`brew install libxcb` 下载的xcb-proto是1.16，libxcb是1.15
1. https://xcb.freedesktop.org/dist/ 下载xcb-proto 1.15，编译安装或者自己创建本地formula安装
2. 执行 `brew link xcb-proto@1.15`在这之前先卸载掉xcb-proto 1.16
3. 执行 `brew edit libxcb` 修改依赖的xcb-proto 的版本，保存退出
    `depends_on "xcb-proto" => :build` 改为 `depends_on "xcb-proto@1.15" => :build`   
4. 执行 `HOMEBREW_NO_INSTALL_FROM_API=1 brew install libxcb`

**Reply by wenma1993** [reply]
我在编译 xcb-proto 1.15的时候 报错 
make[2]: Nothing to be done for `install-exec-am'.
您这边遇到了吗

**Reply by chujiu** [reply]
非常抱歉，这个没有遇到，configure的时候有错误吗
